src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,298 @@
+                        else
+                        {
+                            return InvertIfStyle.WithElseClause;
+                        }
+                    }
+                }
+                else
+                {
+                    if (ifBodyEndPointIsReachable)
+                    {
+                        if (subsequenceSingleExitPointOpt != null &&
+                            SingleSubsequenceStatement(ifStatement))
+                        {
+                            return InvertIfStyle.WithSubsequenceExitPoint;
+                        }
+                        else
+                        {
+                            return InvertIfStyle.MoveIfBodyToElseClause;
+                        }
+                    }
+                    else
+                    {
+                        if (SubsequentStatementsAreInTheSameBlock(ifStatement))
+                        {
+                            return InvertIfStyle.SwapIfBodyWithSubsequence;
+                        }
+                        else
+                        {
+                            return InvertIfStyle.MoveIfBodyToElseClause;
+                        }
+                    }
+                }
+            }
+
+            private bool IsEmptySubsequence(TIfStatementSyntax ifStatement)
+            {
+                foreach (var range in GetSubsequentStatementRange(ifStatement))
+                {
+                    if (!IsEmptyStatementRange(range))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private bool IsEmptyIfBody(TIfStatementSyntax ifStatement)
+            {
+                return IsEmptyStatementRange(GetIfBodyStatementRange(ifStatement));
+            }
+
+            private bool SingleIfBodyStatement(TIfStatementSyntax ifStatement)
+            {
+                var range = GetIfBodyStatementRange(ifStatement);
+                return range.first == range.last;
+            }
+
+            private bool SingleSubsequenceStatement(TIfStatementSyntax ifStatement)
+            {
+                using (var e = GetSubsequentStatementRange(ifStatement).GetEnumerator())
+                {
+                    return e.MoveNext() && e.Current.first == e.Current.last && !e.MoveNext();
+                }
+            }
+
+            private async Task<Document> InvertIfAsync(
+                Document document,
+                TIfStatementSyntax ifStatement,
+                InvertIfStyle invertIfStyle,
+                SyntaxNode subsequenceSingleExitPointOpt,
+                CancellationToken cancellationToken)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                return document.WithSyntaxRoot(
+                    // this returns the root because VB requires changing context around if statement
```

Comment:
note: as per my conversation with @chborl  i think the case that VB is handling is incredibly esoteric.  I would be ok with having VB just checking for that case and *not* offerring invert-if if it helps simplify the rest of the algorithms here.  I think it's a practically irrelevant case in practice and i would be fine with no invert-if for it.

Reply:
(resolved)